### PR TITLE
dai: atomic trigger operations

### PR DESCRIPTION
**Pull request by slawblauciak**
This change adds support for atomic trigger operations,
by grouping DAIs together with a common 'group_id' in
the configuration of each DAI.

When DAIs are in the same group and in the same DSP core,
they will process triggers atomically, i.e. wait for all
the other DAIs to receive the trigger command and pospone
the final, actual trigger operation.

Signed-off-by: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>

### CODE REVIEW

File: src/audio/dai.c [hunk 1]
```diff
@@ -39,6 +39,15 @@ static const struct comp_driver comp_dai;
 DECLARE_SOF_UUID("dai", dai_comp_uuid, 0xc2b00d27, 0xffbc, 0x4150,
 		 0xa5, 0x1a, 0x24, 0x5c, 0x79, 0xc5, 0xe5, 0x4b);
 
+struct dai_group {
+	uint8_t group_id;
+	int num_dais;
+	int trigger_counter;
+	int trigger_cmd;
+	int trigger_ret;
```
Comment (plbossart) [inline src/audio/dai.c:47]: maybe add a comment to explain what those three members represent?

File: src/audio/dai.c [hunk 2]
```diff
@@ -64,6 +74,47 @@ struct dai_data {
...
+static struct dai_group *dai_add_group(struct comp_dev *dev, uint8_t group_id)
+{
+	struct list_item *dai_groups = &sof_get()->dai_groups;
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct list_item *group_item;
+	struct dai_group *group;
+
+	if (dd->group) {
+		comp_info(dev, "dai_add_group(), DAI already in group");
+		return dd->group;
```
Comment (plbossart) [inline src/audio/dai.c:88]: shouldn't this be an error? This would happen e.g. if the config IPC and hence the topology file provide bad data. Better to flag an error than silently go ahead and do things that were not intended.

File: src/audio/dai.c [hunk 3]
```diff
@@ -64,6 +74,47 @@ struct dai_data {
...
+	}
+
+	list_for_item(group_item, dai_groups) {
+		group = container_of(group_item, struct dai_group, list);
+		if (group && group->group_id == group_id)
```
Comment (plbossart) [inline src/audio/dai.c:93]: can group be NULL here? using container_of will return a non-zero value.

File: src/audio/dai.c [hunk 4]
```diff
@@ -64,6 +74,47 @@ struct dai_data {
...
+			break;
+		else
+			group = NULL;
+	}
+
+	if (!group) {
+		group = rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
+				sizeof(*group));
```
Comment (plbossart) [inline src/audio/dai.c:101]: test if allocation failed?

File: src/audio/dai.c [hunk 5]
```diff
@@ -64,6 +74,47 @@ struct dai_data {
...
+
+		list_item_prepend(&group->list, dai_groups);
+		group->group_id = group_id;
+		group->trigger_cmd = -1;
```
Comment (plbossart) [inline src/audio/dai.c:105]: maybe explain what this magic -1 means? or use a define for INVALID or something?

File: src/audio/dai.c [hunk 6]
```diff
@@ -64,6 +74,47 @@ struct dai_data {
...
+	}
+
+	group->num_dais++;
```
Comment (plbossart) [inline src/audio/dai.c:108]: is there any risk of concurrency that might impact the list insertion and num_dais increment? should this be protected with a lock?

File: src/audio/dai.c [hunk 7]
```diff
@@ -635,6 +691,57 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
...
+
+		if (!group->trigger_counter) {
+			irq_local_disable(irq_flags);
+			notifier_event(group, NOTIFIER_ID_DAI_TRIGGER,
+				       BIT(cpu_get_id()), NULL, 0);
+			irq_local_enable(irq_flags);
+
+			ret = group->trigger_ret;
+			group->trigger_cmd = -1;
+			group->trigger_counter = 0;
```
Comment (plbossart) [inline src/audio/dai.c:728]: the trigger_counter is already zero in this if block, is this the right value or is it necessary at all?

File: src/audio/dai.c [hunk 8]
```diff
@@ -635,6 +691,57 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
...
+	return ret;
+}
+
+static void dai_atomic_trigger(void *arg, enum notify_id type, void *data)
+{
+	struct comp_dev *dev = arg;
+	struct dai_data *dd = comp_get_drvdata(dev);
+	int ret = dai_comp_trigger_int(dev, dd->group->trigger_cmd);
+
+	if (ret && !dd->group->trigger_ret)
```
Comment (plbossart) [inline src/audio/dai.c:741]: i didn't get the logic here, what is this trying to test?

File: src/audio/dai.c (hunk 5 above)
Comment (paulstelian97) [reply src/audio/dai.c:105]: Perhaps the proper define name is a namespaced NONE define.

**Comment by slawblauciak** [comment]
@paulstelian97 it doesn't have to perfom any checks, DAIs are always grouped together on a per-core basis, the lists are separate for each SOF instance. And the notifications are registered with appropriate core masks.

The use case is ALH channel aggregation, where we have one stream split into multiple physical links. In such a scenario, we might require the DAIs to be in sync with each other.

File: src/audio/dai.c (hunk 6 above)
Comment (paulstelian97) [reply src/audio/dai.c:108]: Is topology loading concurrent? Do we _load new components_ concurrently? Or component loading is already serialized, even if it's in parallel with active playback/capture?

**Comment by paulstelian97** [comment]
> @paulstelian97 it doesn't have to perfom any checks, DAIs are always grouped together on a per-core basis, the lists are separate for each SOF instance. And the notifications are registered with appropriate core masks.
> 
> The use case is ALH channel aggregation, where we have one stream split into multiple physical links. In such a scenario, we might require the DAIs to be in sync with each other.

What if we receive an invalid topology which somehow doesn't group them correctly in the same core? Or is it syntactically impossible to have a corruption of this kind?

**Comment by slawblauciak** [comment]
@paulstelian97 even if you have them use the same group ID, if they are on different cores, they will still end up being in different groups.

**Comment by paulstelian97** [comment]
> @paulstelian97 even if you have them use the same group ID, if they are on different cores, they will still end up being in different groups.

So groups are really identified by a combination of group ID and core ID? Okay that's fine and that might address Pierre's concern in terms of concurrency as well.

**Comment by slawblauciak** [comment]
@paulstelian97 correct

**Comment by paulstelian97** [comment]
Do add that concern (ALH channel aggregation) in the commit description and we're good.

**Comment by plbossart** [comment]
> > @paulstelian97 even if you have them use the same group ID, if they are on different cores, they will still end up being in different groups.
> 
> So groups are really identified by a combination of group ID and core ID? Okay that's fine and that might address Pierre's concern in terms of concurrency as well.

how is the core information passed? Through the DAI config? we'd really need to document if the group_id is core specific, I didn't click on this when looking at the code.

**Comment by slawblauciak** [comment]
@plbossart pretty sure core ID is in the pipeline config

File: src/audio/dai.c (hunk 3 above)
Comment (slawblauciak) [reply src/audio/dai.c:93]: Yeah, it's just a habit of mine to perform such sanity checks :)

File: src/audio/dai.c (hunk 8 above)
Comment (slawblauciak) [reply src/audio/dai.c:741]: It keeps the first reported trigger error. If there's already an error stored in the field, it won't overwrite it.

Comment (slawblauciak) [reply src/audio/dai.c:741]: I guess I'm gonna change it to last reported error instead. It shouldn't matter and it will let me remove that if statement.

**Comment by plbossart** [comment]
> @plbossart pretty sure core ID is in the pipeline config

it's in the DAI_ADD macro but not DAI_CONFIG.

File: src/audio/dai.c [hunk 9]
```diff
@@ -192,6 +243,11 @@ static void dai_free(struct comp_dev *dev)
 {
 	struct dai_data *dd = comp_get_drvdata(dev);
 
+	if (dd->group) {
+		notifier_unregister(dev, dd->group, NOTIFIER_ID_DAI_TRIGGER);
+		dd->group->num_dais--;
```
Comment (ranj063) [inline src/audio/dai.c:248]: ALso need to delete from the group list?

File: src/audio/dai.c [hunk 10]
```diff
@@ -635,6 +691,57 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
...
+		/* already processing a different trigger command */
+		ret = -EAGAIN;
+	} else {
+		group->trigger_counter--;
+		comp_dbg(dev, "dai_comp_trigger(), trigger counter %d, group %d",
+			 group->trigger_counter, group->group_id);
+
+		if (!group->trigger_counter) {
+			irq_local_disable(irq_flags);
+			notifier_event(group, NOTIFIER_ID_DAI_TRIGGER,
```
Comment (ranj063) [inline src/audio/dai.c:722]: a comment would be good to add here saying that all DAI's in the group will be triggered after the last DAI in the group has received the trigger cmd

File: src/audio/dai.c [hunk 11]
```diff
@@ -635,6 +691,57 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
...
+	}
+
+	return ret;
+}
+
+static void dai_atomic_trigger(void *arg, enum notify_id type, void *data)
+{
+	struct comp_dev *dev = arg;
+	struct dai_data *dd = comp_get_drvdata(dev);
+	int ret = dai_comp_trigger_int(dev, dd->group->trigger_cmd);
```
Comment (ranj063) [inline src/audio/dai.c:739]: This doesnt seem correct to me. You get the notifier after all the last DAI has received the command, so shouldnt we be iterating through the list of all DAI's in the group and calling dai_comp_trigger_int() on all of them?

File: src/include/ipc/dai.h [hunk 12]
```diff
@@ -71,7 +71,8 @@ struct sof_ipc_dai_config {
 
 	/* physical protocol and clocking */
 	uint16_t format;	/**< SOF_DAI_FMT_ */
-	uint16_t reserved16;	/**< alignment */
+	uint8_t group_id;	/**< DAI group ID */
```
Comment (ranj063) [inline src/include/ipc/dai.h:74]: Need a comment here saying that group ID 0 is reserved and means that the DAI doesnt belong to any group

File: src/audio/dai.c (hunk 2 above)
Comment (ranj063) [reply src/audio/dai.c:88]: @plbossart the DAI config IPC is sent more than once especially in the case of HDA DAIs. So I'm not sure this should be an error.

Comment (plbossart) [reply src/audio/dai.c:88]: we could still check that the group_id is the same as configured previously.

File: src/audio/dai.c (hunk 3 above)
Comment (plbossart) [reply src/audio/dai.c:93]: I am also on the paranoid side

File: src/audio/dai.c [hunk 13]
```diff
@@ -557,12 +613,12 @@ static void dai_update_start_position(struct comp_dev *dev)
 }
 
 /* used to pass standard and bespoke command (with data) to component */
-static int dai_comp_trigger(struct comp_dev *dev, int cmd)
+static int dai_comp_trigger_int(struct comp_dev *dev, int cmd)
```
Comment (ranj063) [inline src/audio/dai.c:616]: what does int stand for here?

File: src/audio/dai.c (hunk 9 above)
Comment (slawblauciak) [reply src/audio/dai.c:248]: The group list doesn't store any pointers to DAIs. Only number is stored.

File: src/audio/dai.c (hunk 11 above)
Comment (slawblauciak) [reply src/audio/dai.c:739]: Each DAI is a receiver for the trigger notification, which is called after all of them receive the trigger command. There's no need to iterate here.

File: src/audio/dai.c (hunk 13 above)
Comment (slawblauciak) [reply src/audio/dai.c:616]: Internal

File: src/audio/dai.c (hunk 4 above)
Comment (slawblauciak) [reply src/audio/dai.c:101]: If it fails it will result in immediate panic, since it's the system heap.

File: src/audio/dai.c (hunk 6 above)
Comment (slawblauciak) [reply src/audio/dai.c:127]: I'm not sure I'm afraid, good question.

Comment (paulstelian97) [reply src/audio/dai.c:127]: I'm a bit confused by a thing: You claimed groups are identified by group ID and by core, yet you allocate each group ID only once (and the structure for all groups with the same ID is shared). Shouldn't you have separate structures with separate counters? Or you can just... assume a group ID maps to a single core number?

Comment (paulstelian97) [reply src/audio/dai.c:127]: Blocking on this inconsistency until explained, sorry.

Comment (paulstelian97) [reply src/audio/dai.c:127]: sof_get() returns a single global sof structure. You access a single shared groups list. You search in this shared list and only check group ID. I don't see where the separation between cores comes in, does it come in kernel side? Does the kernel itself ensure that you cannot get a group ID matching for DAIs on multiple cores?

Comment (slawblauciak) [reply src/audio/dai.c:127]: Ah, you're right. I actually thought the structure was separate for each core, but it's not. Alright then

File: src/audio/dai.c (hunk 13 above)
Comment (paulstelian97) [reply src/audio/dai.c:637]: Might be worth fully spelling it out in the function name.

File: src/audio/dai.c (hunk 4 above)
Comment (plbossart) [reply src/audio/dai.c:120]: ack

**Comment by slawblauciak** [comment]
Changes applied. Moved the DAI group list to a core-specific context structure.
I didn't want to have it shared between cores in order to keep the code more readable,
I'd need to introduce spin locks and shared memory handling otherwise.

File: src/audio/dai.c [hunk 14]
```diff
@@ -637,6 +713,62 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
...
+			notifier_event(group, NOTIFIER_ID_DAI_TRIGGER,
+				       BIT(cpu_get_id()), NULL, 0);
+			irq_local_enable(irq_flags);
+
+			ret = group->trigger_ret;
+			group->trigger_cmd = -1; /* no trigger command */
+		}
+	}
+
+	return ret;
```
Comment (paulstelian97) [inline src/audio/dai.c:760]: I see you toy around with the assumption that trigger commands return 0 on success and <0 on failure. Valid assumption, nevertheless it's good to have it documented.

File: src/init/init.c [hunk 15]
```diff
@@ -30,6 +33,20 @@ struct sof *sof_get(void)
 	return &sof;
 }
 
+struct sof_core_context **sof_get_core_context(void)
+{
+	struct core_context *ctx = (struct core_context *)cpu_read_threadptr();
+
+	return &ctx->sof;
+}
```
Comment (paulstelian97) [inline src/init/init.c:41]: This function and related may be worth separating into another commit.

Comment (paulstelian97) [reply src/init/init.c:41]: Also, what happens when slave cores get turned off? Is it possible for that to happen outside the situation that the DSP itself is fully powered off?

File: src/audio/dai.c (hunk 14 above)
Comment (slawblauciak) [reply src/audio/dai.c:760]: I don't quite understand what you mean.
I'm merely forwarding the return value of last trigger operation.

File: src/init/init.c (hunk 15 above)
Comment (slawblauciak) [reply src/init/init.c:41]: I don't know exactly, I'm not familiar with it.
However, judging from the code I investigated, the core context is simply gone in a case like that. So, after the slave core gets turned on again, it allocates the context all over again.

File: src/audio/dai.c (hunk 14 above)
Comment (paulstelian97) [reply src/audio/dai.c:760]: Well in the cores that don't quite trigger you just count the number of triggers and return 0. If the trigger operation itself returns something nonzero on success from DAIs (I'm not sure this is actually the case) then you may break the interface for those that actually use the return value further than the mere error-checking. The fact that you only report error for one of the calls is fine though, because that error will tear down everything anyway and it doesn't matter exactly where it happened.

File: src/init/init.c (hunk 15 above)
Comment (paulstelian97) [reply src/init/init.c:41]: Alright, that _may_ be a leak unless you use the SYSTEM heap for that core (I'm not sure, didn't look carefully enough to check this specifically). I'll leave it up to you to look at that though and say it's okay from my end. Thanks for the explanations.

Comment (slawblauciak) [reply src/init/init.c:41]: Yeah, the system heap is used for that.

File: src/arch/xtensa/xtos/xtos-structs.h [hunk 16]
```diff
@@ -55,6 +56,7 @@ struct core_context {
 	struct schedulers *schedulers;
 	struct notify *notify;
 	struct idc *idc;
+	struct sof_core_context *sof;
```
Comment (tlauda) [inline src/arch/xtensa/xtos/xtos-structs.h:59]: I don't quite understand this change. Why do we add another core context to core context?

Comment (slawblauciak) [reply src/arch/xtensa/xtos/xtos-structs.h:59]: I didn't feel quite right to have audio/component/pipeline-related things in here, so I wanted to move it to a separate structure.

Comment (plbossart) [reply src/arch/xtensa/xtos/xtos-structs.h:59]: it's probably a separate patch anyways?

**Comment by slawblauciak** [comment]
Changes applied.

File: src/include/sof/sof.h [hunk 17]
```diff
@@ -104,4 +104,15 @@ struct sof {
 
 struct sof *sof_get(void);
 
+/**
+ * \brief Core-specific firmware context.
+ */
+struct sof_core_context {
+	/* temporary field to inflate the struct, to be removed afterwards */
+	uint32_t dummy;
```
Comment (lgirdwood) [inline src/include/sof/sof.h]: Still needed ?

File: src/init/init.c [hunk 18]
```diff
@@ -20,7 +21,9 @@
 #include <sof/schedule/task.h>
 #include <sof/sof.h>
 #include <sof/trace/trace.h>
+#include <ipc/topology.h>
 #include <ipc/trace.h>
+#include <xtos-structs.h>
```
Comment (lgirdwood) [inline src/init/init.c:26]: This is an architecture only header so cant be included here, can only be included under arch/xtensa/

File: src/include/sof/sof.h (hunk 17 above)
Comment (slawblauciak) [reply src/include/sof/sof.h]: Still?

Comment (slawblauciak) [reply src/include/sof/sof.h]: The dummy variable is being removed in the next commit in this PR if that's what you mean

**Comment by slawblauciak** [comment]
Changes applied.

File: src/audio/dai.c [hunk 19]
```diff
@@ -41,6 +41,32 @@ DECLARE_SOF_UUID("dai", dai_comp_uuid, 0xc2b00d27, 0xffbc, 0x4150,
...
 
+struct dai_group_list {
+	struct list_item list;
+} __aligned(PLATFORM_DCACHE_ALIGN);
+
+static struct dai_group_list groups[PLATFORM_CORE_COUNT];
+
+struct dai_group {
+	/* Group ID */
+	uint8_t group_id;
```
Comment (lgirdwood) [inline src/audio/dai.c:52]: please use an uint32 here. xtensa generates more code when dealing with uint8

File: src/audio/dai.c [hunk 20]
```diff
@@ -66,6 +93,55 @@ struct dai_data {
 	uint64_t wallclock;	/* wall clock at stream start */
 };
 
+static void dai_atomic_trigger(void *arg, enum notify_id type, void *data);
+
+static int dai_add_group(struct comp_dev *dev, uint8_t group_id)
```
Comment (lgirdwood) [inline src/audio/dai.c:98]: an inline comment would be nice here, also please use a uint32

File: src/audio/dai.c [hunk 21]
```diff
@@ -66,6 +93,55 @@ struct dai_data {
...
+		if (dd->group->group_id != group_id) {
+			comp_err(dev, "dai_add_group(), DAI already in group");
+			return -EINVAL;
+		}
+
+		/* No need to re-assign to the same group, do nothing */
+		return 0;
+	}
+
+	list_for_item(group_item, dai_groups) {
```
Comment (lgirdwood) [inline src/audio/dai.c:115]: what are we looking for here.

File: src/audio/dai.c [hunk 22]
```diff
@@ -66,6 +93,55 @@ struct dai_data {
...
+		group = container_of(group_item, struct dai_group, list);
+
+		if (group->group_id == group_id)
+			break;
+
+		group = NULL;
+	}
+
+	if (!group) {
```
Comment (lgirdwood) [inline src/audio/dai.c:124]: a comment to say our action here.

File: src/audio/dai.c [hunk 23]
```diff
@@ -66,6 +93,55 @@ struct dai_data {
...
+		group = rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
+				sizeof(*group));
+
+		list_item_prepend(&group->list, dai_groups);
+		group->group_id = group_id;
+		group->trigger_cmd = -1; /* no trigger command */
```
Comment (lgirdwood) [inline src/audio/dai.c:130]: Is this the current trigger_state for the group ? if so, I would rename trigger_state and should't the default be 0  (or is 0 used as a state already) ?

File: src/audio/dai.c [hunk 24]
```diff
@@ -66,6 +93,55 @@ struct dai_data {
...
+	}
+
+	group->num_dais++;
+	dd->group = group;
+
+	comp_dbg(dev, "dai_add_group(), group %d num %d",
+		 group_id, group->num_dais);
+
+	notifier_register(dev, group, NOTIFIER_ID_DAI_TRIGGER,
```
Comment (lgirdwood) [inline src/audio/dai.c:139]: an inline comment stating what this notifier is used for

File: src/audio/dai.c (hunk 13 above)
Comment (lgirdwood) [reply src/audio/dai.c:643]: ack, I read it as the type.

File: src/audio/dai.c [hunk 25]
```diff
@@ -637,6 +718,63 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
 	return ret;
 }
 
+static int dai_comp_trigger(struct comp_dev *dev, int cmd)
```
Comment (lgirdwood) [inline src/audio/dai.c:721]: this func needs some inline comments describing each stage

File: src/audio/dai.c [hunk 26]
```diff
@@ -637,6 +718,63 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
...
+			/* return error of last trigger */
+			ret = group->trigger_ret;
+			group->trigger_cmd = -1; /* no trigger command */
+		}
+	}
+
+	return ret;
+}
+
+static void dai_atomic_trigger(void *arg, enum notify_id type, void *data)
```
Comment (lgirdwood) [inline src/audio/dai.c:769]: inline comment on who is holding the locks

File: src/audio/dai.c [hunk 27]
```diff
@@ -917,6 +1062,11 @@ static void sys_comp_dai_init(void)
 {
 	comp_register(platform_shared_get(&comp_dai_info,
 					  sizeof(comp_dai_info)));
+
+	if (cpu_get_id() == PLATFORM_MASTER_CORE_ID)
```
Comment (lgirdwood) [inline src/audio/dai.c:1066]: inline comment on why we only do this for master core.

**Comment by lgirdwood** [comment]
@slawblauciak btw, do you have a kernel PR for the ABI update ?
@kv2019i fyi.

File: src/audio/dai.c (hunk 21 above)
Comment (slawblauciak) [reply src/audio/dai.c:115]: We're checking if a group of a given ID already exists.

File: src/audio/dai.c (hunk 23 above)
Comment (slawblauciak) [reply src/audio/dai.c:130]: It's not state, it's the command to be propagated. 0 can't be default, since it's a valid trigger command.

File: src/audio/dai.c (hunk 26 above)
Comment (slawblauciak) [reply src/audio/dai.c:769]: There are no locks

File: src/audio/dai.c (hunk 24 above)
Comment (paulstelian97) [reply src/audio/dai.c:139]: Used to actually run the command on all components in the group when the last request came through.

File: src/audio/dai.c (hunk 21 above)
Comment (lgirdwood) [reply src/audio/dai.c:115]: ok, so it's best to comment this for readability, I would recommend this for all list_for_each searches now.

File: src/audio/dai.c (hunk 23 above)
Comment (lgirdwood) [reply src/audio/dai.c:130]: ah, ok - it seems like we need a enum/macro for the -1 cmd here.

File: src/audio/dai.c (hunk 26 above)
Comment (lgirdwood) [reply src/audio/dai.c:769]: ok, but who is expected to set the atomic context ?

Comment (slawblauciak) [reply src/audio/dai.c:769]: Ah, yes, you're right. The context is set by the last DAI to receive the trigger command. I shall mention it in a comment.

**Comment by slawblauciak** [comment]
Requested changes applied.
Kernel-side PR hasn't been made yet. Still need a minor ABI version assigned.

File: src/audio/dai.c (hunk 23 above)
Comment (slawblauciak) [reply src/audio/dai.c]: Very well, added COMP_TRIGGER_NONE

File: src/audio/dai.c [hunk 28]
```diff
@@ -66,6 +93,59 @@ struct dai_data {
 	uint64_t wallclock;	/* wall clock at stream start */
 };
 
+static void dai_atomic_trigger(void *arg, enum notify_id type, void *data);
+
+static int dai_add_group(struct comp_dev *dev, uint32_t group_id)
```
Comment (mmaka1) [inline src/audio/dai.c:98]: Should be named dai_add_to_group()? Actually this function does both, creates(adds) a group and adds dai to the group. So how about creating dai_group API for group mgmt similar to the one for dai mgmt defined in src/lib/dai.c? So that dai_groups in sof is manipulated via that API? dai_group_get(..., flag = DG_CREAT /* create if does not exist *?), dai_group_add() etc. ?

File: src/audio/dai.c [hunk 29]
```diff
@@ -66,6 +93,59 @@ struct dai_data {
...
+{
+	int cpu_id = cpu_get_id();
+	struct list_item *dai_groups = &sof_get()->dai_groups[cpu_id].list;
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct list_item *group_item;
+	struct dai_group *group;
+
+	if (dd->group) {
+		if (dd->group->group_id != group_id) {
+			comp_err(dev, "dai_add_group(), DAI already in group");
```
Comment (mmaka1) [inline src/audio/dai.c:108]: It is good when error message delivers as much debug information as possible, i.e. target group_id and assigned group id could be printed.

File: src/audio/dai.c [hunk 30]
```diff
@@ -66,6 +93,59 @@ struct dai_data {
...
+
+		if (group->group_id == group_id)
+			break;
+
+		group = NULL;
+	}
+
+	/* Group doesn't exist, let's allocate and initialize it */
+	if (!group) {
+		group = rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
```
Comment (mmaka1) [inline src/audio/dai.c:128]: This is another reason for dai_group API in lib. dai component code (application layer) should not allocate from the SYS zone.

**Comment by slawblauciak** [comment]
Changes applied.

**Comment by slawblauciak** [comment]
SOFCI TEST

**Comment by slawblauciak** [comment]
Some weird testbench error, no idea what to make of it

File: src/audio/dai.c [hunk 31]
```diff
@@ -51,6 +51,7 @@ struct dai_data {
 	struct timestamp_cfg ts_config;
 	struct dai *dai;
 	struct dma *dma;
+	struct dai_group *group;
```
Comment (lgirdwood) [inline src/audio/dai.c:54]: Does NULL mean no group ? if so , best to have a comment here stating such.

File: src/audio/dai.c [hunk 32]
```diff
@@ -66,6 +67,42 @@ struct dai_data {
 	uint64_t wallclock;	/* wall clock at stream start */
 };
 
+static void dai_atomic_trigger(void *arg, enum notify_id type, void *data);
+
+static int dai_assign_group(struct comp_dev *dev, uint32_t group_id)
```
Comment (lgirdwood) [inline src/audio/dai.c:72]: an inline comment describing what this does.

File: src/audio/dai.c [hunk 33]
```diff
@@ -637,6 +679,70 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
...
 }
 
+static int dai_comp_trigger(struct comp_dev *dev, int cmd)
+{
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct dai_group *group = dd->group;
+	uint32_t irq_flags;
+	int ret = 0;
+
+	/* DAI not in a group */
```
Comment (lgirdwood) [inline src/audio/dai.c:689]: /* use normal trigger if DAI is not grouped */

File: src/audio/dai.c [hunk 34]
```diff
@@ -637,6 +679,70 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
...
+	if (!group) {
+		comp_dbg(dev, "dai_comp_trigger(), non-atomic trigger");
+		return dai_comp_trigger_internal(dev, cmd);
+	}
+
+	if (!group->trigger_counter) {
```
Comment (lgirdwood) [inline src/audio/dai.c:695]: /* DAI is grouped so only trigger when final DAI is ready */

File: src/audio/dai.c [hunk 35]
```diff
@@ -740,6 +847,12 @@ static int dai_config(struct comp_dev *dev, struct sof_ipc_dai_config *config)
 		return -EINVAL;
 	}
 
+	if (config->group_id)
+		ret = dai_assign_group(dev, config->group_id);
+
```
Comment (lgirdwood) [inline src/audio/dai.c:852]: newline not needed here.

**Comment by lgirdwood** [comment]
New pattern on CI, lets reurn to be sure we are goo, could be DUT

**Comment by lgirdwood** [comment]
SOFCI TEST

File: src/audio/dai.c [hunk 36]
```diff
@@ -66,6 +67,43 @@ struct dai_data {
...
+	}
+
+	dd->group = dai_group_get(group_id, DAI_CREAT);
+	if (!dd->group) {
+		comp_err(dev, "dai_assign_group(), failed to assign group %d",
+			 group_id);
+		return -EINVAL;
+	}
+
+	dd->group->num_dais++;
```
Comment (mmaka1) [inline src/audio/dai.c:95]: Might be better to encapsulate the counter operations in dai_group_get()/put(). The caller, called by dai_free(), could also clear the group_id setting it to 0 if num_dais goes down to 0. Then the group object could be reused later.

**Comment by lgirdwood** [comment]
@kv2019i are you guys good with this for kernel ?
@slawblauciak any update ?

**Comment by plbossart** [comment]
@lgirdwood we are missing the new topology token for group_id and the ABI update to this to be used by the kernel

**Comment by slawblauciak** [comment]
I've been putting off the kernel side ABI change, I'll get it done soon, probably this week.

**Comment by slawblauciak** [comment]
Kernel-side PR: https://github.com/thesofproject/linux/pull/2331

**Comment by lgirdwood** [comment]
@slawblauciak ping

**Comment by slawblauciak** [comment]
Seems like some test randomly failed, let me rerun

**Comment by slawblauciak** [comment]
SOFCI TEST

File: src/include/ipc/dai.h [hunk 37]
```diff
@@ -71,7 +71,8 @@ struct sof_ipc_dai_config {
 
 	/* physical protocol and clocking */
 	uint16_t format;	/**< SOF_DAI_FMT_ */
-	uint16_t reserved16;	/**< alignment */
+	uint8_t group_id;	/**< DAI group ID, 0 means no group */
+	uint8_t reserved8;	/**< alignment */
```
Comment (kv2019i) [inline src/include/ipc/dai.h:75]: @lgirdwood A general question: should we mandate that ABI version is mentioned in doxygen for any new ABI struct fields? Now you basicly have to dig through firmware git logs to figure out whether a certain field was always there, or added later.

Comment (lgirdwood) [reply src/include/ipc/dai.h:75]: @kv2019i yes, sounds like a good ide. Btw is this requirement in sof-docs ?
@slawblauciak care to add the comment - I think we are good to go if kernel is good  ?

**Comment by slawblauciak** [comment]
Rebased and applied changes.
Refactored the group API a bit and introduced re-use of released DAI groups, as per @mmaka1's suggestion

**Comment by slawblauciak** [comment]
Alrighty, tests passed.

**Comment by lgirdwood** [comment]
@slawblauciak shall we rerun CI ? Maybe a DUT or CI server needs rebooted ?

**Comment by slawblauciak** [comment]
@lgirdwood supposed network issues, just asked ZiBi about it, he said he'll re-run that one.

**Comment by lgirdwood** [comment]
Jenkins showing known and unrelated issues.

**Comment by lgirdwood** [comment]
@kv2019i are you guys now good from kernel ?

**Comment by plbossart** [comment]
> Alrighty, tests passed.

@slawblauciak can you point us to the tests used? was this limited to firmware tests (I don't see a topology token so not sure how this would be tested at the kernel level).

**Comment by plbossart** [comment]
> @kv2019i are you guys now good from kernel ?

I'd like something at the kernel level showing how this can be tested (maybe this is just a test topology that's needed with a group_id token?)

**Comment by slawblauciak** [comment]
@plbossart they were Python FW tests with the slim driver.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
rerun CI to pick up new BSW+ ICL  fixes.

**Comment by plbossart** [comment]
> rerun CI to pick up new BSW+ ICL fixes.

@lgirdwood This is all good really. the ICL part is not fully fixed, we need to fix the test scripts: the error is because the 'panic' word is detected but without checking the error level. We'll have smarter filters soon that will only use the err and warn levels to report issues.
